In [1]:
from collections import Counter
from imblearn.over_sampling import SMOTE
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import metrics
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("../../data/sea/all_merged_dataset_with_tfidf.csv")

In [3]:
x = df.loc[:, df.columns!='valid_patent'].astype(str)
y = df['valid_patent'].astype(int)
print(x.shape, y.shape)

(1905, 24719) (1905,)


In [4]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=50)

In [5]:
print('Original dataset shape {}'.format(Counter(y_train)))
sm = SMOTE(random_state=42)
x_res , y_res = sm.fit_sample(x_train, y_train)
print('Resampled dataset shape {}'.format(Counter(y_res)))

Original dataset shape Counter({0: 1311, 1: 22})
Resampled dataset shape Counter({0: 1311, 1: 1311})


In [6]:
patent_tree = tree.DecisionTreeClassifier(criterion='gini', splitter='random', max_depth=5)
patent_tree.fit(x_res, y_res)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=5,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='random')

In [7]:
y_pred_tr = patent_tree.predict(x_test)
print('Accuracy: %.2f' % metrics.accuracy_score(y_test, y_pred_tr))

Accuracy: 0.96


In [8]:
# target_names = ['class 0', 'class 1', 'class 2']
print(metrics.classification_report(y_test, y_pred_tr))

             precision    recall  f1-score   support

          0       1.00      0.96      0.98       565
          1       0.19      0.71      0.30         7

avg / total       0.99      0.96      0.97       572



             precision    recall  f1-score   support

          0       0.99      0.91      0.95       565
          1       0.06      0.43      0.10         7

avg / total       0.98      0.91      0.94       572

Accuracy: 0.91


            precision    recall  f1-score   support

          0       1.00      0.96      0.98       565
          1       0.17      0.71      0.28         7

avg / total       0.99      0.95      0.97       572

Accuracy: 0.95